lib/parsers/gjs-parser.js

```diff
@@ -0,0 +1,430 @@
+        n.head.range = [
+          range[0] + docLines.positionToOffset(n.head.loc.start),
+          range[0] + docLines.positionToOffset(n.head.loc.end),
+        ];
+        n.head.loc = {
+          start: codeLines.offsetToPosition(n.head.range[0]),
+          end: codeLines.offsetToPosition(n.head.range[1]),
+        };
+      }
+      n.range =
+        n.type === 'Template'
+          ? [tpl.replacedRange[0], tpl.replacedRange[1]]
+          : [
+              range[0] + docLines.positionToOffset(n.loc.start),
+              range[0] + docLines.positionToOffset(n.loc.end),
+            ];
+
+      n.start = n.range[0];
+      n.end = n.range[1];
+      n.loc = {
+        start: codeLines.offsetToPosition(n.range[0]),
+        end: codeLines.offsetToPosition(n.range[1]),
+      };
+      if (n.type === 'Template') {
+        n.loc.start = codeLines.offsetToPosition(tpl.templateRange[0]);
+        n.loc.end = codeLines.offsetToPosition(tpl.templateRange[1]);
+      }
+      if ('blockParams' in n) {
+        n.params = [];
+      }
+      if ('blockParams' in n && n.parent) {
+        let part = code.slice(...n.parent.range);
+        let start = n.parent.range[0];
+        let idx = part.indexOf('|') + 1;
+        start += idx;
+        part = part.slice(idx, -1);
+        idx = part.indexOf('|');
+        part = part.slice(0, idx);
+        for (const param of n.blockParams) {
+          const regex = new RegExp(`\\b${param}\\b`);
+          const match = part.match(regex);
+          const range = [start + match.index, 0];
+          range[1] = range[0] + param.length;
+          n.params.push({
+            type: 'BlockParam',
+            name: param,
+            range,
+            parent: n,
+            loc: {
+              start: codeLines.offsetToPosition(range[0]),
+              end: codeLines.offsetToPosition(range[1]),
+            },
+          });
+        }
+      }
+      n.type = `Glimmer${n.type}`;
+      allNodeTypes.add(n.type);
+    }
+    // tokens should not contain tokens of comments
+    ast.tokens = ast.tokens.filter(
+      (t) => !comments.some((c) => c.range[0] <= t.range[0] && c.range[1] >= t.range[1])
+    );
+    ast.contents = template;
+    tpl.ast = ast;
+  }
+  for (const [k, v] of Object.entries(glimmerVisitorKeys)) {
+    templateVisitorKeys[`Glimmer${k}`] = [...v];
+  }
+  return {
+    templateVisitorKeys,
+    templateInfos,
+    comments,
+  };
+}
+
+/**
+ * traverses the AST and replaces the transformed template parts with the Glimmer
+ * AST.
+ * This also creates the scopes for the Glimmer Blocks and registers the block params
+ * in the scope, and also any usages of variables in path expressions
+ * this allows the basic eslint rules no-undef and no-unsused to work also for the
+ * templates without needing any custom rules
+ * @param result
+ * @param preprocessedResult
+ * @param visitorKeys
+ */
+function convertAst(result, preprocessedResult, visitorKeys) {
```

Comment:
why is all this code needed? 
do we need to zip in the glimmer AST with the JS AST?
does doing so allow us to eventually phase out template-lint? (or adopt the rules in to here?)

Reply:
yes, possibly. but mostly now to auto-detect unused and undef vars